src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultTagHelperActivator.cs

```diff
@@ -2,79 +2,45 @@
-        private readonly Func<Type, PropertyActivator<ViewContext>[]> _getPropertiesToActivate;
+        private readonly ITypeActivatorCache _typeActivatorCache;
 
         /// <summary>
         /// Instantiates a new <see cref="DefaultTagHelperActivator"/> instance.
         /// </summary>
-        public DefaultTagHelperActivator()
+        /// <param name="typeActivatorCache">The <see cref="ITypeActivatorCache"/>.</param>
+        public DefaultTagHelperActivator(ITypeActivatorCache typeActivatorCache)
         {
-            _injectActions = new ConcurrentDictionary<Type, PropertyActivator<ViewContext>[]>();
-            _getPropertiesToActivate = type =>
-                PropertyActivator<ViewContext>.GetPropertiesToActivate(
-                    type,
-                    typeof(ViewContextAttribute),
-                    CreateActivateInfo);
-        }
-
-        /// <inheritdoc />
-        public void Activate<TTagHelper>(TTagHelper tagHelper, ViewContext context)
-            where TTagHelper : ITagHelper
-        {
-            if (tagHelper == null)
-            {
-                throw new ArgumentNullException(nameof(tagHelper));
-            }
-
-            if (context == null)
-            {
-                throw new ArgumentNullException(nameof(context));
-            }
-
-            var propertiesToActivate = _injectActions.GetOrAdd(
-                tagHelper.GetType(),
-                _getPropertiesToActivate);
-
-            for (var i = 0; i < propertiesToActivate.Length; i++)
+            if (typeActivatorCache == null)
             {
-                var activateInfo = propertiesToActivate[i];
-                activateInfo.Activate(tagHelper, context);
+                throw new ArgumentNullException(nameof(typeActivatorCache));
             }
 
-            InitializeTagHelper(tagHelper, context);
+            _typeActivatorCache = typeActivatorCache;
         }
 
-        private static void InitializeTagHelper<TTagHelper>(TTagHelper tagHelper, ViewContext context)
+        /// <inheritdoc />
+        public TTagHelper Create<TTagHelper>(ViewContext context)
             where TTagHelper : ITagHelper
         {
-            // Run any tag helper initializers in the container
-            var serviceProvider = context.HttpContext.RequestServices;
-            var initializers = serviceProvider.GetService<IEnumerable<ITagHelperInitializer<TTagHelper>>>();
-
-            foreach (var initializer in initializers)
+            if (context == null)
             {
-                initializer.Initialize(tagHelper, context);
+                throw new ArgumentNullException(nameof(context));
             }
-        }
 
-        private static PropertyActivator<ViewContext> CreateActivateInfo(PropertyInfo property)
-        {
-            return new PropertyActivator<ViewContext>(property, viewContext => viewContext);
+            return _typeActivatorCache.CreateInstance<TTagHelper>(
```

Comment:
Referring to the code in the controller activator - that's currently on my list to review as a perf hotspot. 

I think it's fine to ask for forgiveness rather than permission here, and I wish we had done the same in the controller activator.